tests/sqlx/lexer.spec.ts

```diff
@@ -3,69 +3,145 @@ import { expect } from "chai";
+        },
+        {
+          in: "pre_operations {select regexp_extract('01a_data_engine', '^(\\\\d{2}\\\\w)')}",
+          expected: "select regexp_extract('01a_data_engine', '^(\\\\\\\\d{2}\\\\\\\\w)')"
+        },
+        {
+          in: "pre_operations {select * from regexp_extract('\\\\', '')}",
+          expected: "select * from regexp_extract('\\\\\\\\', '')"
+        },
+        { in: "pre_operations {}", expected: "" }
       ];
-      pairs.forEach(pair => {
-        expect(parseSqlx(pair[0]).preOperations).eql([pair[1]]);
+      tests.forEach(test => {
+        expect(parseSqlx(test.in).preOperations).eql([test.expected]);
       });
     });
     it("nothing in a string is interpreted as a special term", () => {
-      const pairs = [
-        ['post_operations {select "asd\'123"}', 'select "asd\'123"'],
-        ["post_operations {select 'asd\"123'}", "select 'asd\"123'"],
-        [
-          "post_operations {select * from regexp_extract('js {', \"\")}",
-          "select * from regexp_extract('js {', \"\")"
-        ]
+      const tests: Array<{ in: string; expected: string }> = [
+        { in: 'post_operations {select "asd\'123"}', expected: 'select "asd\'123"' },
+        { in: "post_operations {select 'asd\"123'}", expected: "select 'asd\"123'" },
+        {
+          in: "post_operations {select * from regexp_extract('js {', \"\")}",
+          expected: "select * from regexp_extract('js {', \"\")"
+        }
       ];
-      pairs.forEach(pair => {
-        expect(parseSqlx(pair[0]).postOperations).eql([pair[1]]);
+      tests.forEach(test => {
+        expect(parseSqlx(test.in).postOperations).eql([test.expected]);
       });
     });
   });
+  describe("syntax tree construction", () => {
+    it("strings don't affect the tree", () => {
+      const tree = constructSyntaxTree("SELECT SUM(IF(track.event = 'example', 1, 0)) js { }");
+      const expected = {
+        contentType: "sql",
+        contents: [
+          "SELECT SUM(IF(track.event = 'example', 1, 0)) ",
+          { contentType: "js", contents: ["js { }"] }
+        ]
+      };
+      expect(tree).eql(expected);
+    });
+    it("inline js blocks tokenized", () => {
+      const tree = constructSyntaxTree("select * from ${ref('dab')}");
+      const expected = {
+        contentType: "sql",
+        contents: ["select * from ", { contentType: "jsPlaceholder", contents: ["${ref('dab')}"] }]
+      };
+      expect(tree).eql(expected);
+    });
+    it("inline js blocks tokenized correctly if string present beforehand", () => {
+      const tree = constructSyntaxTree('select regexp("", ${ref("dab")})');
+      const expected = {
+        contentType: "sql",
+        contents: [
+          'select regexp("", ',
+          { contentType: "jsPlaceholder", contents: ['${ref("dab")}'] },
+          ")"
+        ]
+      };
+      expect(tree).eql(expected);
+    });
+    it("inline js blocks tokenized correctly", () => {
```

Comment:
this is a bit of an epic test, seems like it might be testing too much (it was useful for debugging, but not necessarily something we want to maintain in a test). can we boil it down to something more specific + smaller?

Reply:
Good point, have migrated the important string bit to the test above.